modyn/supervisor/internal/triggers/performancetrigger.py

```diff
@@ -0,0 +1,302 @@
+
+        # allows to detect drift in a fixed interval
+        self._sample_left_until_detection = config.detection_interval_data_points
+
+        self.data_density = DataDensityTracker(config.data_density_window_size)
+        self.performance_tracker = PerformanceTracker(config.performance_triggers_window_size)
+
+        self.decision_policies = _setup_decision_policies(config)
+
+        self._triggered_once = False
+        self.model_refresh_needed = False
+        self._metrics = setup_metrics(config.evaluation.dataset.metrics)
+        self._last_detection_interval: list[tuple[int, int]] = []
+
+        self._leftover_data: list[tuple[int, int]] = []
+        """Stores data that was not processed in the last inform call because
+        the detection interval was not filled."""
+
+        self._label_transformer_function = (
+            deserialize_function(
+                config.evaluation.label_transformer_function,
+                LABEL_TRANSFORMER_FUNC_NAME,
+            )
+            if config.evaluation.label_transformer_function
+            else None
+        )
+
+    def init_trigger(self, context: TriggerContext) -> None:
+        self.context = context
+        self._init_dataloader_info()
+        self._init_model_downloader()
+
+    def inform(
+        self,
+        new_data: list[tuple[int, int, int]],
+        log: TriggerPolicyEvaluationLog | None = None,
+    ) -> Generator[int, None, None]:
+        # pylint: disable=too-many-locals
+        new_key_ts = self._leftover_data + [(key, timestamp) for key, timestamp, _ in new_data]
+        # reappending the leftover data to the new data requires incrementing the sample left until detection
+        self._sample_left_until_detection += len(self._leftover_data)
+
+        # index of the first unprocessed data point in the batch
+        processing_head_in_batch = 0
+
+        # Go through remaining data in new data in batches of `detect_interval`
+        while True:
+            if self._sample_left_until_detection - len(new_key_ts) > 0:
+                # No detection in this trigger because of too few data points to fill detection interval
+                self._leftover_data = new_key_ts
+                self._sample_left_until_detection -= len(new_key_ts)
+                return
+
+            # At least one detection, fill up window up to that detection
+            next_detection_interval = new_key_ts[: self._sample_left_until_detection]
+
+            # Update the remaining data
+            processing_head_in_batch += len(next_detection_interval)
+            new_key_ts = new_key_ts[len(next_detection_interval) :]
+
+            # Reset for next detection
+            self._sample_left_until_detection = self.config.detection_interval_data_points
+
+            # Run the evaluation (even if we don't use the result, e.g. for the first forced trigger)
+            self.data_density.inform_data(next_detection_interval)
```

Comment:
Do we always need to keep track of data density? In my understanding, we only need it when forecasting, don't we?

Reply:
Yes, we need it iff at least one of the `DecisionCriterion` needs it. Otherwise, it's unused. However, it's quite cheap to keep track of this, so I'd rather always do it for less code complexity.

Reply:
Okay, I was just thinking about the memory requirements of this, but as long as it's not a problem, we can leave it as is